File: codes/java/1044.最长重复子串.java
import java.util.HashMap;

class Solution {
	public String longestDupSubstring(String s) {
		int l = 0;
		int r = s.length() - 1;
		// 二分查找子串，如果子串是重复子串，那么把长度加大再查找，如果不是重复子串，那么把长度减小再查找
		while (l < r) {
			int m = l + ((r - l + 1) >> 1);
			if (isDuplicatePresent(s, m)) {
				l = m;
			} else {
				r = m - 1;
			}
		}
		return findDuplicate(s, l);
	}

	// 查找末尾为length的重复子串，如果找到了，返回子串，否则返回空串
	private String findDuplicate(String s, int length) {
		long hash = 0; // 子串哈希值
		int prime = 29; // 哈希计算的质数，当每个字符都是小写字母时，取29最好
		long firstEntryPower = 1; // 用于计算子串哈希值的第一项的幂
		HashMap<Long, Integer> map = new HashMap<>();

		char[] sArr = s.toCharArray();

		for (int i = 0; i < length; i++) {
			firstEntryPower *= prime;
			hash = hash * prime + (sArr[i] - 'a');
		}
		map.put(hash, 0);
		for (int i = length; i < s.length(); i++) {
			hash = hash * prime + sArr[i] - 'a';
			hash -= firstEntryPower * (sArr[i - length] - 'a');

			if (map.containsKey(hash)) {
				int idx = map.get(hash);
				return s.substring(idx, idx + length);
			}
			map.put(hash, i - length + 1);
		}
		return "";
	}

	private boolean isDuplicatePresent(String s, int length) {
		if (length == 0) {
			return true;
		}
		return !findDuplicate(s, length).isEmpty();
	}
}

File: codes/python/1044.最长重复子串.py


class Solution:
	def longestDupSubstring(self, s: str) -> str:
		l = 0
		r = len(s) - 1
		while l < r:
			m = l + ((r - l + 1) >> 1)
			if self.isDuplicatePresent(s, m):
				l = m
			else:
				r = m - 1
		return self.findDuplicate(s, l)
	
	def findDuplicate(self, s: str, length: int) -> str:
		prime = 29 # 选取一个质数，因为质数的乘法比较少，所以哈希冲突的概率比较小，这里选择29的原因是因为26个字母加上3个特殊字符，正好是29个
		hash = 0 # 初始化哈希值为0
		power = 1 # 初始化权重为1
		map = {}
		sArr = list(s)
		for i in range(length):
			power *= prime
			hash = hash * prime + (ord(sArr[i]) - ord('a'))
		map[hash] = 0
		for i in range(length, len(s)):
			hash = hash * prime + ord(sArr[i]) - ord('a') - power * (ord(sArr[i - length]) - ord('a'))
			if hash in map:
				idx = map[hash]
				return s[idx:idx + length]
			map[hash] = i - length + 1
		return ""

	def isDuplicatePresent(self, s: str, length: int) -> bool:
		if length == 0:
			return True
		return self.findDuplicate(s, length) != ""